<h1>1. Print, Calculator, Variable Types, and Operators</h1>
<h2>10/01/20</h2>

<h2>1.0 Introduction to Jupyter Notebooks</h2>
First, this is a Jupyter Notebook, which enables a combination of markup cells (like this one) to provide description alongside actual code that can be run. If you want to change a markup cell, simply double-click on the text, make any changes you want to make, and click "Run" <b>above</b> to execute. You can add HTML-style tags to markup cells to add details such as formatting and links.

There are also code cells, which can actually be executed, just like in a normal programming environment. Click on the below code cell and try clicking "Run" (or type Shift+Enter). You can also modify the text that's being printed to the screen and try running it again.

In [1]:
print("Hello World!")

Hello World!


<h2>1.1 Basic Variables</h2>
Python has <b>dynamic typing</b>, which means that variable types are set automatically based on the input. This is handy if you're coming from an environment like Fortran that requires you to specify each and every time what kind of variable you want to create!

Here are some of the most important variable types (more on these in a bit):
<ul>
    <li>Integer (short and long)</li>
    <li>Floating point (float)</li>
    <li>Strings</li>
    <li>Booleans</li>
    <li>NoneType</li>
    <li>Lists and tuples</li> (Monday's class)
    <li>Dictionaries</li> (Monday's class)
</ul>

Arithmetic operators:
<ul>
    <li>+ is addition</li>
    <li>- is subtraction</li>
    <li>/ is division</li>
    <li>* is multiplication</li>
    <li>** is exponentiation</li>
</ul>

Comparison operators:
<ul>
    <li>> is greater than</li>
    <li>< is less than</li>
    <li>>= is greater than or equal to</li>
    <li>&lt= is less than or equal to</li>
    <li>!= is not equal to</li>
    <li>== is equal to</li>
</ul>

<b>Important note:</b> Python is case-sensitive! "A" is not the same thing as "a".

In [13]:
# Code that is preceded with a hashtag will not be read by the computer.
# These comments are intended for human readers of this code!
# Commenting your code is essential to ensure that you and others
# will be able to interpret it down the road!

# Now, let's try playing with some numbers...

a = 3.6
print(a)
type(a)

3.6


float

In [5]:
b = -2.1
print(b)
type(b)

-2.1


float

In [8]:
c = 3
print(c)
type(c)

3


int

In [11]:
d = -4
print(d)
type(d)

-4


int

Note that all of the above are examples of Python's <b>dynamic typing</b>: assigning a variable a floating point value (i.e., including a decimal point) results in that variable being a float, whereas assigning a variable an integer value (i.e., no decimal point) results in that variable being an integer.

You can also convert from one to the other using the int() and float() commands!

In [14]:
# Convert a from a float to an integer.

print(int(a))

3


In [15]:
# Convert c from an integer to a float.

print(float(c))

3.0


Note that the conversion from float to int <b>loses information</b> along the way: Python merely truncates everything after the decimal point.

On the other hand, the conversion from int to float is <b>essentially lossless</b>: the same information is maintained.

In [16]:
# Now let's try some mathematical operators.
print(a)
print(b)
print(c)
print(d)

3.6
-2.1
3
-4


In [17]:
print(a*b)

-7.5600000000000005


In [18]:
print(b+c)

0.8999999999999999


In [19]:
print(a/c)

1.2


In [21]:
print(d/c)

-1.3333333333333333


When applying mathematical operations, Python tries to keep as much information as possible! This means that an integer and a float interacting together will result in a float as an answer. Likewise, division of integers will give you a float by default (this is new in version 3 of Python).

<b>Why is a*b not exactly -7.56?</b>

Floats are not exact numbers! There will always be some degree of rounding errors associated with floating point numbers (because we don't have infinite storage to keep an infinite number of digits after the decimal). As a result, it's good practice when comparing floats not to look for equality, but instead to look for "close enough" - there are functions to do this that we will see later!

<h2>1.2 Strings</h2>

Strings are the text variable type in Python: they're surrounded either by single or double quotes (it generally doesn't matter which, as long as you're consistent at the start and finish).

In [22]:
e = 'hello'
print(e)
type(e)

hello


str

If you want to combine two strings, you can do so easily with the "+" command.

In [23]:
f = 'there'
g = e+f
print(g)

hellothere


In [24]:
# To add a space, simply throw another string in there.

g = e+' '+f
print(g)

hello there


In [27]:
# How about combining a string and a float?

h = 2
i = 'meters'
g = h+' '+i
print(g)

TypeError: unsupported operand type(s) for +: 'int' and 'str'

The <b>TypeError</b> you get in the above example explains the problem: a float cannot be combined in this way with a string. But what if we want to make this change? That's where the str() function comes in handy.

In [28]:
# Let's try that again!

g = str(h)+' '+i
print(g)

2 meters


Sometimes you want a new line within the same string. This can be accomplished with the "escape character", which in Python is a backslash: \\. "\n" will start a new line!

In [29]:
# Creating a multi-line string using the newline character.

j = "This string has two lines. Here's the first..."+"\n"+"...and here's the second."
print(j)

This string has two lines. Here's the first...
...and here's the second.


<h2>1.3 Boolean Variables</h2>

A Boolean variable can only have two values: <b>True</b> and <b>False</b> (and remember, capitalization counts!) These are the results of the logical tests mentioned earlier!

In [32]:
i = 1
j = 1
k = 2

m = (i == j)
print(m)
type(m)

True


bool

In [33]:
print(k <= i)

False


In [34]:
print(j != i)

False


You can also combine conditions using <b>and</b> and <b>or</b>.

<b>and</b>: Both statements must be <b>True</b> in order for the entire statement to be <b>True</b>.

<b>or</b>: Either statement (or both) can be <b>True</b> in order for the entire statement to be <b>True</b>.

In [38]:
print(True and False)

False


In [39]:
print(True or False)

True


In [ ]:
print(i == j and k == j)

In [36]:
print(i == j or k == j)

True


<h2>1.4 NoneType</h2>

A NoneType variable can only have one value: <b>None</b> (and capitalization counts!). 

Why would we want this? Consider the scenario where you want to make sure there's no value already in a given variable (so, for example, if you're working with thousands of lines of code and the same variable types keep coming up). Setting that variable to None (via something like <b>a = None</b>) means you're guaranteed not to be messing around with any previous values by accident.

<h1>TAKE-HOME POINTS</h1>

<ul>
    <li>Python has a variety of data types, many of which we saw today!</li>
    <li>An <b>integer</b> variable is a number with no decimal point.</li>
    <li>A <b>floating point</b> variable is a number with a decimal point</li>
    <li>A <b>string</b> variable is text.</li>
    <li>A <b>Boolean</b> variable can take on the values <b>True</b> and <b>False</b>.</li>
    <li>A <b>NoneType</b> variable can take on the value <b>None</b>.</li>
    <li>Python uses dynamic typing, which means you don't need to define a variable's type when you first use that variable.</li>
    <li>The type() command lets you see what type a given variable is. int(), float(), and str() can be used to convert to different variable types.</li>
    <li>Combining a float and an integer results in a float.</li>
    <li>Combining a string and an integer or float results in a TypeError.</li>
    <li>Floats are often prone to rounding errors.</li>
    <li><b>and</b> and <b>or</b> can be used to combine Boolean variables.</li>
</ul>